File: content/examples/other/leetcode/problems/decode-ways/example-2/Solution.java
public class Solution {
  static int numDecodings(String s) {
    if (s.length() == 0) return 0;
    int dp2 = 1;
    int dp1 = s.charAt(0) == '0' ? 0 : 1;
    for (int i = 2; i <= s.length(); i += 1) {
      int cur = 0;
      if (s.charAt(i - 1) != '0') cur += dp1;
      int two = Integer.parseInt(s.substring(i - 2, i));
      if (two >= 10 && two <= 26) cur += dp2;
      dp2 = dp1;
      dp1 = cur;
    }
    return dp1;
  }

  public static void main(String[] args) {
    System.out.println(numDecodings("226"));
  }
}

File: content/examples/other/leetcode/problems/decode-ways/example-2/solution.py
def num_decodings(s):
    if not s:
        return 0
    dp2 = 1
    dp1 = 0 if s[0] == "0" else 1
    for i in range(2, len(s) + 1):
        cur = 0
        if s[i - 1] != "0":
            cur += dp1
        two = int(s[i - 2:i])
        if 10 <= two <= 26:
            cur += dp2
        dp2, dp1 = dp1, cur
    return dp1


if __name__ == "__main__":
    print(num_decodings("226"))
